sim/tools/runner.ts

```diff
@@ -129,3 +139,72 @@ class RawBattleStream extends BattleStreams.BattleStream {
 		super._write(message);
 	}
 }
+
+class DualStream {
+	private debug: boolean;
+	private readonly control: RawBattleStream;
+	private test: RawBattleStream;
+
+	constructor(input: boolean, debug: boolean) {
+		this.debug = debug;
+		// The input to both streams should be the same, so to satisfy the
+		// input flag we only need to track the raw input of one stream.
+		this.control = new RawBattleStream(input);
+		this.test = new RawBattleStream(false);
+	}
+
+	get rawInputLog() {
+		const control = this.control.rawInputLog;
+		const test = this.test.rawInputLog;
+		assert.deepStrictEqual(test, control);
+		return control;
+	}
+
+	async read() {
+		const control = await this.control.read();
+		const test = await this.test.read();
+		// In debug mode, wait to catch this as a difference in the inputLog
+		// and error there so we get the full battle state dumped instead.
+		if (!this.debug) assert.strictEqual(test, control);
+		return control;
+	}
+
+	write(message: string) {
+		this.control._write(message);
+		this.test._write(message);
+		this.compare();
+	}
+
+	async end() {
+		// We need to compare first because _end() destroys the battle object.
+		this.compare(true);
+		await this.control._end();
+		await this.test._end();
+	}
+
+	compare(end?: boolean) {
+		if (!this.control.battle || !this.test.battle) return;
+
+		const control = this.control.battle.toJSON();
+		const test = this.test.battle.toJSON();
+		try {
+			assert.deepStrictEqual(test, control);
+		} catch (err) {
+			if (this.debug) {
+				// NOTE: diffing these directly won't work because the key ordering isn't stable.
+				fs.writeFileSync('logs/control.json', JSON.stringify(control, null, 2));
+				fs.writeFileSync('logs/test.json', JSON.stringify(test, null, 2));
+			}
+			throw new Error(err.message);
+		}
+
+		if (end) return;
+		const send = this.test.battle.send;
+		// NOTE: retaining trunc as part of the round trip is only required for certain custom
+		// game formats which override it in the format onBegin logic to have non-standard behavior.
+		const trunc = this.test.battle.trunc;
```

Comment:
`trunc` is a function and can't be serialized into JSON.

I'd recommend just defining `trunc` directly into the format so it can be manually imported.

Reply:
Obviously. `trunc` isn't being serialized here - we're saving it *before* we round trip and reinstalling it afterwards.

> I'd recommend just defining `trunc` directly into the format so it can be manually imported.

???

Reply:
> ???

```diff
-onBegin() { this.trunc = n => Math.trunc(n); },
+trunc(n) { return Math.trunc(n); },
```

Add some glue code for replacing `trunc` with the version in the format, at the start of a battle, and away we go.

Reply:
Done.